SwiftIntro.playground/Pages/Untitled Page.xcplaygroundpage/Contents.swift

```diff
@@ -134,48 +174,66 @@ class ColorPickerVC {
 class AddItemViewController {
     var itemUploadForm: ItemUploadForm = ItemUploadForm()
     var colorPicker = ColorPickerVC()
-
+    
     func openColorPicker() {
-        //print("present color picker")
+        print("present color picker")
         
         //give value for ColorPickerVC onComplete closure. String received from closure should be added to ItemUploadForm as colorId
-        //colorPicker.onComplete =
+        colorPicker.onComplete = { [weak self] color in
+            self?.itemUploadForm.colorId = color
+            print(color)
+        }
         
         //Simulating color selection in colorPickerVC
         colorPicker.selectRedColor()
-        //print("Color was selected")
+        print("Color was selected")
+    }
+    
+    deinit {
+        print("AddItemViewController deinitialized")
     }
 }
 
 var addItemVC: AddItemViewController? = AddItemViewController()
 addItemVC?.openColorPicker()
+addItemVC = nil
 
 //Does it have a memory leak? prove it and fix it, does your code above has a memory leak?
 
 class DownloaderManager {
-    var imagesDownloader: ImagesDownloader?
+    weak var imagesDownloader: ImagesDownloader?
     
     func doManagerThings() {
-        //print("Do manager things")
+        print("Do manager things")
+    }
+    
+    deinit {
+        print("DownloaderManager was deinitialized")
     }
 }
 
 class ImagesDownloader {
     var onPublish: ((_ value: String) -> Void)?
     var images: [String] = [String]()
     
-    var downloadManager: DownloaderManager?
+    weak var downloadManager: DownloaderManager?
     
     init() {
-        onPublish = { imgName in
-            self.images.append(imgName)
-            //print("new img appended, img count: \(self.images.count)")
+        onPublish = { [weak self] imgName in
+            if let safeSelf = self {
+                safeSelf.images.append(imgName)
+                print("new img appended, img count: \(safeSelf.images.count)")
+            }
```

Comment:
Nice! My recommendation, use guards in these places, so there will be less nesting. We could use guard let as a check for self, and after this check self is safe to use
```suggestion
            guard let self = self else {
                return
            }
            safeSelf.images.append(imgName)
            print("new img appended, img count: \(safeSelf.images.count)")
```